### ISSUE TITLE
Can't drain ReplicaSet pods with 1.2

### ISSUE THREAD

**Issue by maclof** [kubernetes/kubernetes#23531]
Whenever I run the kubectl drain command, I get an output similar to the following (the throttling request messages) and the node isn't drained of pods:

```bash
root@node4:~/kubernetes/virtual-machines# kubectl drain node7.web.ahead4.com --grace-period=300
node "node7.web.ahead4.com" already cordoned
W0327 12:40:23.064522   20276 request.go:627] Throttling request took 140.826987ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/fords-coaches
W0327 12:40:23.264460   20276 request.go:627] Throttling request took 195.885379ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/ftd-crm
W0327 12:40:23.464502   20276 request.go:627] Throttling request took 196.130639ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/icspa
W0327 12:40:23.664448   20276 request.go:627] Throttling request took 195.78121ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/insta-uk-crm
W0327 12:40:23.864451   20276 request.go:627] Throttling request took 196.284324ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/itstrain
W0327 12:40:24.064477   20276 request.go:627] Throttling request took 194.578357ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/oprg
W0327 12:40:24.264519   20276 request.go:627] Throttling request took 195.540545ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/pbssltd
W0327 12:40:24.464607   20276 request.go:627] Throttling request took 196.578202ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/platform-demos-shop
W0327 12:40:24.664516   20276 request.go:627] Throttling request took 195.68612ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/safetysigns-catalog
W0327 12:40:24.864484   20276 request.go:627] Throttling request took 195.852022ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/sealesgrabhire
W0327 12:40:25.064500   20276 request.go:627] Throttling request took 195.808393ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/udf-spectrum
W0327 12:40:25.264513   20276 request.go:627] Throttling request took 195.844258ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/udf-udf
W0327 12:40:25.464498   20276 request.go:627] Throttling request took 195.512148ms, request: GET:http://localhost:8080/api/v1/namespaces/default/replicationcontrollers/william-fountain
error: pods not managed by ReplicationController, Job, or DaemonSet: ahead4-docs-46695868-kdr6c (use --force to override)
```

**Comment by AdoHe** [comment]
```
If there are any pods that are neither mirror pods nor managed--by ReplicationController, DaemonSet or Job--, then drain will not delete any pods unless you use --force.
```
Also the throttling request log has been removed in one of latest PRs.

**Comment by dchen1107** [comment]
@maclof Can you try to use --force to see if your issue is fixed here? Please let us know if you need further support. Thanks!

**Comment by maclof** [comment]
The "ahead4-docs-46695868-kdr6c" pod that it mentioned in the log was created by a ReplicaSet, are these not evicted by using drain?

I ended up having to manually delete each pod on that node after marking it as unschedulable. If I have to use this again I'll try --force next time.

**Comment by a-robinson** [comment]
Huh, if drain doesn't take ReplicaSets into account, that could be a problem. What version of kubectl were you using?

**Comment by maclof** [comment]
@a-robinson I'm using both 1.2.0 for the server and client:
```bash
root@node4:~/kubernetes/ingress-proxy# kubectl version
Client Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.0", GitCommit:"5cb86ee022267586db386f62781338b0483733b3", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.0", GitCommit:"5cb86ee022267586db386f62781338b0483733b3", GitTreeState:"clean"}
```

I have just tested this again:
```bash
root@node4:~/kubernetes/ingress-proxy# kubectl drain node7.web.ahead4.com --grace-period=300
node "node7.web.ahead4.com" cordoned
error: pods not managed by ReplicationController, Job, or DaemonSet: ahead4-docs-46695868-7354w,rbd (use --force to override)
```

If I use the same command with --force I get the following output:
```
root@node4:~/kubernetes/ingress-proxy# kubectl drain node7.web.ahead4.com --grace-period=300 --force
node "node7.web.ahead4.com" already cordoned
WARNING: About to delete these pods not managed by ReplicationController, Job, or DaemonSet: ahead4-docs-46695868-7354w,rbd
pod "ahead4-docs-46695868-7354w" deleted
pod "rbd" deleted
node "node7.web.ahead4.com" drained
```

However if I run `docker ps` directly on the node, I can see that the pod is still there:
```bash
root@node7:~# docker ps
CONTAINER ID        IMAGE                                 COMMAND                  CREATED             STATUS              PORTS               NAMES
90b8a0724ee0        docker.ahead4.com/ahead4/docs:0.0.5   "/opt/init"              5 minutes ago       Up 5 minutes                            k8s_ahead4-docs.7fecd8fd_ahead4-docs-46695868-7354w_default_5a00f187-f70e-11e5-88a9-9e920b2e1578_46ac3730
12e13152d475        gcr.io/google_containers/pause:2.0    "/pause"                 6 minutes ago       Up 6 minutes                            k8s_POD.cf58006d_ahead4-docs-46695868-7354w_default_5a00f187-f70e-11e5-88a9-9e920b2e1578_19d42a92
41e26c4f8b85        solnetcloud/keepalived:latest         "/usr/local/bin/entry"   13 hours ago        Up 13 hours                             keepalived
56aad960356b        maclof/ceph-docker:latest             "/entrypoint.sh mds"     3 days ago          Up 3 days                               ceph-mds
afcfa1c2e17e        maclof/ceph-docker:latest             "/entrypoint.sh osd_c"   3 days ago          Up 40 hours                             ceph-osd
3756957e48b6        30a0d8a371ef                          "/entrypoint.sh mon"     3 days ago          Up 3 days                               ceph-mon
```

The ReplicaSet is as follows:
```bash
root@node4:~/kubernetes/ingress-proxy# kubectl describe rs ahead4-docs-46695868
Name:           ahead4-docs-46695868
Namespace:      default
Image(s):       docker.ahead4.com/ahead4/docs:0.0.5
Selector:       name=ahead4-docs,pod-template-hash=46695868
Labels:         name=ahead4-docs,pod-template-hash=46695868
Replicas:       1 current / 1 desired
Pods Status:    2 Running / 0 Waiting / 0 Succeeded / 0 Failed
Volumes:
  ahead4-docs-secrets:
    Type:       Secret (a volume populated by a Secret)
    SecretName: ahead4-docs-secrets
Events:
  FirstSeen     LastSeen        Count   From                            SubobjectPath   Type            Reason                  Message
  ---------     --------        -----   ----                            -------------   --------        ------                  -------
  8m            8m              1       {replicaset-controller }                        Normal          SuccessfulCreate        Created pod: ahead4-docs-46695868-7354w
  2m            2m              1       {replicaset-controller }                        Normal          SuccessfulCreate        Created pod: ahead4-docs-46695868-lwb7f
```

This is created by a deployment:
```bash
root@node4:~/kubernetes/ingress-proxy# kubectl describe deployment ahead4-docs
Name:                   ahead4-docs
Namespace:              default
CreationTimestamp:      Fri, 25 Mar 2016 15:18:47 +0100
Labels:                 name=ahead4-docs
Selector:               name=ahead4-docs
Replicas:               1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  1 max unavailable, 1 max surge
OldReplicaSets:         <none>
NewReplicaSet:          ahead4-docs-46695868 (1/1 replicas created)
No events.
```

**Comment by maclof** [comment]
Sorry, ignore what I said previously about the pod still being there, it looks like it actually abided by the grace period and has now scheduled the pod onto another node (This is indicated by the 2 running on the RS pod status).

However it definitely doesn't seem to delete the pod if I don't use the --force option.

**Comment by maclof** [comment]
I have submitted a PR to fix this issue: https://github.com/kubernetes/kubernetes/pull/23689

**Comment by davidopp** [comment]
@mml

**Comment by mml** [comment]
Thanks for the fix, @maclof.  Centralizing the code that answers "is this pod managed by a controller?" would help us avoid this.  #2210 will make the check easier, but either way there should be a library call, and creators of new controllers should know they need to update that library.

**Comment by davidopp** [comment]
https://github.com/kubernetes/kubernetes/issues/14541#issuecomment-176427996

**Comment by quinton-hoole** [comment]
Assigned to @mml as per #23689

**Comment by davidopp** [comment]
Fixed in #23689
